### PR TITLE
dependency_collector: fix caching of deps requiring brewed curl

### PR DESCRIPTION
**Pull request by EricFromCanada**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
This addresses a rather subtle dependency bug that's best illustrated with an example. If a formula's source needs brewed curl to be downloaded, it's marked as `using: :homebrew_curl` and will show curl as an implicit dependency.

```console
$ brew deps --include-build --direct --tree --annotate libzip
libzip
├── curl  [build] [test] [implicit]
├── cmake  [build]
└── zstd 
```

Same if the formula is a dependent for another.

```console
$ brew deps --include-build --tree --annotate xlsxio
xlsxio
└── libzip 
    ├── curl  [build] [test] [implicit]
    │   ├── pkg-config  [build]
...
    ├── cmake  [build]
    └── zstd 
        ├── cmake  [build]
        ├── lz4 
        └── xz 
```

But in some cases, curl isn't added as an implicit dependency when it should be.

```console
$ brew deps --include-build --tree --annotate libisofs
libisofs
├── autoconf  [build]
│   └── m4 
├── automake  [build]
│   └── autoconf 
│       └── m4 
├── libtool  [build]
│   └── m4 
└── libzip 
    ├── cmake  [build]
    └── zstd 
        ├── cmake  [build]
        ├── lz4 
        └── xz 
```

In the above example, removing autoconf/automake/libtool will make curl show properly as a libzip dependency. So what's going on? 

In #12296, I included a dirty workaround for the possibility that a resource requiring brewed curl to download (i.e. using `HomebrewCurlDownloadStrategy`) may also require another implicit dependency to open the downloaded file, which was to add curl directly to the dependency list with `@deps << brewed_curl_dep_if_needed(tags)`, and then proceed with analyzing the URL as is done with the regular `CurlDownloadStrategy`. The problem is that Dependency objects are cached. So, if a formula using e.g [libzip](https://formulae.brew.sh/formula/libzip) (which is a .xz file) also depends on another formula which is the same file type (e.g. [libtool](https://formulae.brew.sh/formula/libtool), also a .xz file) and that dependency is processed first, the cache says that no implicit dependencies are required for this file type, causing the above code to be skipped. 

The fix is to 1) use a separate cache key for `HomebrewCurlDownloadStrategy` dependency objects, and 2) instead of inserting into `@deps` directly, send multiple items back to `add()` and insert them there.

### CODE REVIEW

**Review by Bo98 — APPROVED** [review]
Makes sense, thanks!

**Comment by Bo98** [comment]
I wonder if the `Resource` case could be generalised to `spec.download_strategy` (+ `File.extname(spec.url)` in the curl case), given all of the other download strategies look like they could also be cached. Not sure if any speedup would be noticeable

**Comment by EricFromCanada** [comment]
Yep, seems to work. Also added a case for `VCSDownloadStrategy` which didn't exist when the original method was written.

File: Library/Homebrew/dependency_collector.rb [hunk 1]
```diff
@@ -63,11 +65,14 @@ def fetch(spec)
...
 
   def cache_key(spec)
-    if spec.is_a?(Resource) && spec.download_strategy <= CurlDownloadStrategy
-      File.extname(spec.url)
-    else
-      spec
+    if spec.is_a?(Resource)
+      if spec.download_strategy <= CurlDownloadStrategy
+        return "#{spec.download_strategy}#{File.extname(spec.url).split("?").first}"
+      elsif spec.download_strategy <= VCSDownloadStrategy
```
Comment (Bo98) [inline Library/Homebrew/dependency_collector.rb:71]: Do we need the condition at all? Would this work?

```suggestion
      else
```

**Review by Bo98 — APPROVED** [review]
LGTM

File: Library/Homebrew/dependency_collector.rb (hunk 1 above)
Comment (EricFromCanada) [reply Library/Homebrew/dependency_collector.rb:71]: It should if it's a safe assumption that any Resource will have a download strategy defined.

**Comment by MikeMcQuaid** [comment]
Thanks @EricFromCanada, good catch!

**Pull request merged**